### PR TITLE
OCPBUGS-27215: Replace routes from kubevirt live migration to prevent ECMP

### PR DESCRIPTION
**Pull request by qinqon**
**- What this PR does and why is it needed**
Kubevirt live migration between nodes that do not own the VM subnet are
creating wrongly at ECMP route since is not replacing the point to point
routes, this breaks traffic.

This change use just just IPPrefix and Policy at the point to
point routes predicate to prevent ECMP.

closes [#OCPBUGS-27215](https://issues.redhat.com/browse/OCPBUGS-27215)

**- Special notes for reviewers**
This is tested at unit and e2e tests, for unit a new test is added that sets the virt-launcher at a third node that do not own the subnet and create a target virt-launcher at different one not owning the subnet either.

For e2e instead of doing two migrations, where we migrate back to the node owning the subnet, a extra migration is added to migrate between two nodes not owning the subnet.

A e2e migration example with subnbets
- node1 subnet is 192.168.1.0/24
- node2 subnet is 192.168.2.0/24
- node3 subnet is 192.168.3.0/24

Vm is created at node1 so it gets the IP 192.168.1.2
- First migration do a migration from node1 to node2 so it has to create point to point routes for the first time at local and remote zones
- Second migration is done from node2 to node3 so it will have to replace the remote zone ptp routes to local ones since now VM is running there (this is what this PR is fixing)
- Third migration goes back to node1 and since it's running at the original node ptp routes are not needed and normal ovn-k routing is fine.

**- Description for the changelog**
Replace live migration routes to prevent ECMP.

### CODE REVIEW

File: test/e2e/kubevirt.go [hunk 1]
```diff
@@ -81,6 +81,23 @@ passwd:
   - name: core
     password_hash: $y$j9T$b7RFf2LW7MUOiF4RyLHKA0$T.Ap/uzmg8zrTcUNXyXvBvT26UgkC6zZUVg3UKXeEp5
 `
+		labelNode = func(nodeName, label string) error {
```
Comment (oshoval) [inline test/e2e/kubevirt.go:84]: can this cosmetics be split please to another commit ? (also the usage of it of course)

File: go-controller/pkg/kubevirt/router.go [hunk 2]
```diff
@@ -150,7 +150,7 @@ func EnsureLocalZonePodAddressesToNodeRoute(watchFactory *factory.WatchFactory,
 			},
 		}
 		if err := libovsdbops.CreateOrReplaceLogicalRouterStaticRouteWithPredicate(nbClient, types.OVNClusterRouter, &ingressRoute, func(item *nbdb.LogicalRouterStaticRoute) bool {
-			matches := item.IPPrefix == ingressRoute.IPPrefix && item.Nexthop == ingressRoute.Nexthop && item.Policy != nil && *item.Policy == *ingressRoute.Policy
+			matches := item.IPPrefix == ingressRoute.IPPrefix && item.Policy != nil && *item.Policy == *ingressRoute.Policy
```
Comment (maiqueb) [inline go-controller/pkg/kubevirt/router.go:153]: I don't understand why removing the `nextHop` from the predicate fixes the issue. 

I (and other reviewers) would appreciate if you could explain that in the PR's description - it currently says:
> This change use just IPPrefix and Nexthop for the predicate point to point routes are updated and not creating preventing from ECMP routes to be configured.

which is wrong IMO - you remove nexthop from the predicate, thus you can't use it for the predicate.

**Review by maiqueb — CHANGES_REQUESTED** [review]

File: go-controller/pkg/ovn/kubevirt_test.go [hunk 3]
```diff
@@ -1007,6 +1024,71 @@ var _ = Describe("OVN Kubevirt Operations", func() {
 					},
 				},
 			}),
+			Entry("for pre-copy live migration at local zone from remote zone", testData{
```
Comment (oshoval) [inline go-controller/pkg/ovn/kubevirt_test.go:1027]: consider please adding to PR desc how it is tested
why 3 nodes ?
and the flow, according the ticket, it should migrate back and forth to the same node etc
if this what happens here (didnt dive yet)

File: go-controller/pkg/kubevirt/router.go (hunk 2 above)
Comment (qinqon) [reply go-controller/pkg/kubevirt/router.go:153]: Ahh the comment is wrong... I will fix it.

File: go-controller/pkg/ovn/kubevirt_test.go (hunk 3 above)
Comment (qinqon) [reply go-controller/pkg/ovn/kubevirt_test.go:1027]: Done

File: go-controller/pkg/kubevirt/router.go (hunk 2 above)
Comment (qinqon) [reply go-controller/pkg/kubevirt/router.go:153]: Done

File: test/e2e/kubevirt.go (hunk 1 above)
Comment (qinqon) [reply test/e2e/kubevirt.go:84]: is not a super big refactoring, better let's keep it at the same commit.

File: go-controller/pkg/kubevirt/router.go (hunk 2 above)
Comment (maiqueb) [reply go-controller/pkg/kubevirt/router.go:153]: makes sense. since these are p2p routes - /32 or /128 prefix len routes - we can safely replace them.

File: go-controller/pkg/ovn/kubevirt_test.go (hunk 3 above)
Comment (maiqueb) [inline go-controller/pkg/ovn/kubevirt_test.go:1027]: I think we can better describe what is happening here; what if we described the test as `pre-copy live migration between zones that do not own the original subnet` ? 

I think my proposal can still be improved, but IMO it is more descriptive.

File: test/e2e/kubevirt.go [hunk 4]
```diff
@@ -122,31 +139,31 @@ passwd:
...
 				selectedNodes = append(selectedNodes, nodes[0])
-				if len(selectedNodes) == 2 {
-					break // we want just a pair of them
+				if len(selectedNodes) == 3 {
+					break // we want just three of them
 				}
 			}
 		}
 
+		Expect(selectedNodes).To(HaveLen(3), "at least three nodes at different zones incase of interconnect are needed to run tests")
```
Comment (maiqueb) [inline test/e2e/kubevirt.go:160]:
```suggestion
		Expect(selectedNodes).To(HaveLen(3), "at least three nodes in different zones are needed for interconnect scenarios")
```

**Review by maiqueb — APPROVED** [review]
I just have a small test naming nit.

It's hard to visualize the updated e2e test, but unfortunately I don't have a better suggestion on how to improve its readability.

**Comment by jcaamano** [comment]
LGTM

@qinqon let me know when you have considered @maiqueb comments

**Comment by qinqon** [comment]
> LGTM
> 
> @qinqon let me know when you have considered @maiqueb comments

@jcaamano addressed Miguel's comments.

**Comment by jcaamano** [comment]
re-running flaked jobs

**Review by jcaamano — APPROVED** [review]
